pyteal/ast/opup.py

```diff
@@ -0,0 +1,164 @@
+    through the foreign apps array.
+    """
+
+    """The app to call must be provided by the user."""
+    Explicit = 0
+
+    """The app to call is created then deleted for each request to increase budget."""
+    OnCall = 1
+
+
+ON_CALL_APP = Bytes("base16", "06810143")  # v6 pyteal program "Int(1)"
+MIN_TXN_FEE = Int(1000)
+
+
+class OpUp:
+    """Utility for increasing opcode budget during app execution.
+    Note: the available budget must be high enough to execute the budget
+    increase code prior to invoking ensure_budget() or maximize_budget().
+
+    Example:
+        .. code-block:: python
+
+            opup = OpUp(OpUpMode.OnCall)
+            program_with_opup = Seq(
+                ...,
+                opup.ensure_budget(Int(1000)),
+                ...,
+            )
+    """
+
+    def __init__(self, mode: OpUpMode, target_app_id: Expr = None):
+        if mode == OpUpMode.Explicit:
+            if target_app_id is None:
+                raise TealInputError(
+                    "target_app_id must be specified in Explicit OpUp mode"
+                )
+            require_type(target_app_id, TealType.uint64)
+            self.target_app_id = target_app_id
+        elif mode == OpUpMode.OnCall:
+            self.target_app_id_slot = ScratchSlot()
+        else:
+            raise TealInputError("Invalid OpUp mode provided")
+
+        self.mode = mode
+
+        # A budget buffer is necessary to deal with an edge case of ensure_budget():
+        #   if the current budget is equal to or only slightly higher than the
+        #   required budget then it's possible for ensure_budget() to return with a
+        #   current budget less than the required budget. The buffer prevents this
+        #   from being the case.
+        self.buffer = Int(50)
+
+    def _create_app(self) -> Expr:
+        self.target_app_id = self.target_app_id_slot.load()
+        return Seq(
+            InnerTxnBuilder.Begin(),
+            InnerTxnBuilder.SetFields(
+                {
+                    TxnField.type_enum: TxnType.ApplicationCall,
+                    TxnField.approval_program: ON_CALL_APP,
+                    TxnField.clear_state_program: ON_CALL_APP,
+                }
+            ),
+            InnerTxnBuilder.Submit(),
+            self.target_app_id_slot.store(Gitxn[0].created_application_id()),
+        )
+
+    def _delete_app(self, app_id: Expr) -> Expr:
+        return Seq(
+            InnerTxnBuilder.Begin(),
+            InnerTxnBuilder.SetFields(
+                {
+                    TxnField.type_enum: TxnType.ApplicationCall,
+                    TxnField.on_completion: OnComplete.DeleteApplication,
+                    TxnField.application_id: app_id,
+                }
+            ),
+            InnerTxnBuilder.Submit(),
+        )
+
+    def _ensure_budget_expr(self, required_budget: Expr) -> Expr:
+        return While(required_budget < Global.opcode_budget()).Do(
```

Comment:
@algoidurovic Minor:  Is the intend inequality `<` or `<=`?

I raise the question because I thought it'd be `<=`.  When `=`, I thought there's no need to acquire more opcode budget because the compared budget reflects the budget.

If there's a change to be made, note that there's 2 places to make the change.

Reply:
> When =, I thought there's no need to acquire more opcode budget because the compared budget reflects the budget.

If I'm understanding you correctly, the existing comparison does enforce this. `<=` would continue the loop when the values being compared are equal.

Reply:
@algoidurovic Duh - Sorry for the noise here.  Disregard - resolving.